snapshots/windows/windows.go

```diff
@@ -391,3 +396,99 @@ func (s *snapshotter) parentIDsToParentPaths(parentIDs []string) []string {
 	}
 	return parentLayerPaths
 }
+
+// This is essentially a recreation of what HCS' CreateSandboxLayer does with some extra bells and
+// whistles like expanding the volume if a size is specified. This will create a 1GB scratch
+// vhdx to be used if a different sized scratch that is not equal to the default of 20 is requested.
+func (s *snapshotter) createScratchLayer(ctx context.Context, snDir string, parentLayers []string, sizeGB int) error {
+	parentLen := len(parentLayers)
+	if parentLen == 0 {
+		return errors.New("no parent layers present")
+	}
+	baseLayer := parentLayers[parentLen-1]
+
+	var (
+		templateBase     = filepath.Join(baseLayer, "blank-base.vhdx")
+		templateDiffDisk = filepath.Join(baseLayer, "blank.vhdx")
+		expand           = sizeGB > 0 && sizeGB != 20
+	)
+
+	// If a size greater than 0 or not equal to 20 (the default size produced by hcs)
+	// was specified we make a new set of disks to be used. We make it a 1GB disk and just
+	// expand it to the size specified so for future container runs we don't need to remake a disk.
+	if expand {
+		templateBase = filepath.Join(baseLayer, "scratch.vhdx")
+		templateDiffDisk = filepath.Join(baseLayer, "scratch-diff.vhdx")
+	}
+
+	if _, err := os.Stat(templateDiffDisk); os.IsNotExist(err) {
+		// Scratch disk not present so lets make it.
+		if err := computestorage.SetupContainerBaseLayer(ctx, baseLayer, templateBase, templateDiffDisk, 1); err != nil {
```

Comment:
Doesn't this try to create the base hives & the layout file? And those files will already be present in the layer directory right?

Reply:
Correct, however in `SetupContainerBaseLayer` I remove them if they exist otherwise like you're probably alluding to the call to `SetupBaseOsLayer` will fail. https://github.com/microsoft/hcsshim/blob/fd21b8d1922c7fb8b4a50c76d048fe1a69b7e7dc/computestorage/helpers.go#L38

Reply:
Oh my bad. I was looking a different version of `SetupContainerBaseLayer`.

Reply:
If I understand correctly, using this function on an _existing_ Base layer as is done here, if it fails partway through, will leave the base layer corrupted, as the Hives directory and layout file have been removed. I think I'm seeing this after rebasing the Windows snapshotter support onto this PR, in #4419, as `SetupContainerBaseLayer` is failing in the call to `FormatWritableLayerVhd`.